File: Week_01/cpp/src/_977_SquaresOfASortedArray.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    // 读懂题目的意思了，想到了归并，但是没有逆向思考。
    vector<int> sortedSquares(vector<int> &A)
    {
        int n = A.size(), i = 0, j = n - 1, k = j;
        vector<int> ret(n);
        while (i <= j)
        {
            if (A[i] * A[i] < A[j] * A[j])
            {
                ret[k--] = A[j] * A[j--];
            }
            else
            {
                ret[k--] = A[i] * A[i++];
            }
        }
        return ret;
    }
};

File: Week_01/java/src/_977_SquaresOfASortedArray.java
public class _977_SquaresOfASortedArray {
    class Solution {
        public int[] sortedSquares(int[] A) {
            if (A == null || A.length == 0) return A;
            int n = A.length;
            int[] ret = new int[n];
            int l = 0, r = n - 1, k = r;
            while (l <= r) {
                if (A[l] * A[l] < A[r] * A[r]) {
                    ret[k--] = A[r] * A[r];
                    r--;
                } else {
                    ret[k--] = A[l] * A[l];
                    l++;
                }
            }
            return ret;
        }
    }
}
